sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/providers/JavaRowUdf.java

```diff
@@ -0,0 +1,323 @@
+    if (inputSchema.hasField(expression)) {
+      final int ix = inputSchema.indexOf(expression);
+      return new FunctionAndType(
+          inputSchema.getField(expression).getType(), (Row row) -> row.getValue(ix));
+    } else {
+      Map<String, Type> fieldTypes = new HashMap<>();
+      for (Schema.Field field : inputSchema.getFields()) {
+        if (expression.indexOf(field.getName()) != -1) {
+          fieldTypes.put(field.getName(), typeFromFieldType(field.getType()));
+        }
+      }
+      Type type = StringCompiler.guessExpressionType(expression, fieldTypes);
+      StringBuilder source = new StringBuilder();
+      source.append("import java.util.function.Function;\n");
+      source.append("import " + Row.class.getTypeName() + ";\n");
+      source.append("public class Eval implements Function<Row, Object> {\n");
+      source.append("  public Object apply(Row __row__) {\n");
+      for (Map.Entry<String, Type> fieldEntry : fieldTypes.entrySet()) {
+        source.append(
+            String.format(
+                "    %s %s = (%s) __row__.getValue(%s);%n",
+                fieldEntry.getValue().getTypeName(),
+                fieldEntry.getKey(),
+                fieldEntry.getValue().getTypeName(),
+                inputSchema.indexOf(fieldEntry.getKey())));
+      }
+      source.append("    return " + expression + ";\n");
+      source.append("  }\n");
+      source.append("}\n");
+      return new FunctionAndType(
+          type, (Function<Row, Object>) StringCompiler.getInstance("Eval", source.toString()));
+    }
+  }
+
+  @SuppressWarnings({
+    "nullness" // TODO(https://github.com/apache/beam/issues/20497)
+  })
+  private static FunctionAndType createFuctionFromCallable(String callable)
+      throws StringCompiler.CompileException, ReflectiveOperationException {
+    Matcher matcher =
+        Pattern.compile("\\bpublic\\s+class\\s+(\\S+)", Pattern.MULTILINE).matcher(callable);
+    Preconditions.checkArgument(matcher.find(), "No public class defined in callable source.");
+    return new FunctionAndType(
+        (Function<Row, Object>) StringCompiler.getInstance(matcher.group(1), callable.toString()));
+  }
+
+  @SuppressWarnings({
+    "nullness" // TODO(https://github.com/apache/beam/issues/20497)
+  })
+  private static FunctionAndType createFunctionFromName(String name, String path)
+      throws ReflectiveOperationException, MalformedURLException {
+    ClassLoader classLoader =
+        path == null
+            ? ClassLoader.getSystemClassLoader()
+            : new URLClassLoader(
+                new URL[] {new URL("file://" + path)}, ClassLoader.getSystemClassLoader());
+    String className, methodName = null;
+    if (name.indexOf("::") == -1) {
+      className = name;
+      methodName = null;
+    } else {
+      String[] parts = name.split("::", 2);
+      className = parts[0];
+      methodName = parts[1];
+    }
```

Comment:
@robertwb Shouldn't there be support for reading a file from GCS in this logic by 2.52?

Reply:
I don't see that as as critical for 2.52 (v.s. the basic functionality of supporting expressions), but we could try to get that in as a follow-up.

Reply:
My only concern would be if a user specified a GCS path in their pipeline which uses the python MapToFields, but then updates the pipeline to use more Java transforms, causing the MapToFields to switch to the Java implementation.

Is there a way to force an SDK implementation for a transform?

Reply:
Here the language is explicitly specified; there's no automatic switching between the two. (E.g. a python script is not interchangeable with a java jar).

Reply:
I went ahead and implemented this. Ideally it'd use the artifact service, but we don't yet have a way to declare that from a transform.